internal/topology/check/compatibility.go

```diff
@@ -18,39 +18,179 @@ limitations under the License.
 package check
 
 import (
-	"github.com/pkg/errors"
+	"fmt"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
 // ReferencedObjectsAreStrictlyCompatible checks if two referenced objects are strictly compatible, meaning that
 // they are compatible and the name of the objects do not change.
-func ReferencedObjectsAreStrictlyCompatible(current, desired client.Object) error {
+func ReferencedObjectsAreStrictlyCompatible(current, desired client.Object) field.ErrorList {
```

Comment:
Let's drop Referenced given that we are passing the actual object, so can't really make assumption about the object origin.